### ISSUE TITLE
PsiFixFromReal for large numbeers (> 31 bits) fails on Modelsim

### ISSUE THREAD

**Issue by obruendl** [paulscherrerinstitute/psi_fix#11]
In GHDL the testbench runs fine, in Modelsim the code behaves differently and fails.

**Comment by obruendl** [comment]
I investigated the issue a bit. It looks like modelsim is not even executing the *PsiFixFromReal()* call, so it looks like a modelsim bug.

What I tested:
I added a print (*report "bla"*) inside PsiFixFromReal. Before the failing call of PsiFixFromReal I also added a print (to know which part of the transcript to look at).
I saw the *bla* getting reported for all calls of PsiFixFromReal except for the failing one.

This week I run out of time to hunt the issue but I will continue searching for a workaround.

**Comment by obruendl** [comment]
I found the issue: for operations with mixed and integer arguments GHDL seems to up-convert (from integer to real) while modelsim down-converts (real to integer). This did not work. I will fix it.

**Comment by obruendl** [comment]
I fixed the issue in en_cl_fix. Note that at least version 1.1.6 is required to use the fix.

**Comment by rybaniec-r** [comment]
> 
> 
> I fixed the issue in en_cl_fix. Note that at least version 1.1.6 is required to use the fix.

Now there are no errors in Modelsim. However, there are new problems when using GHDL:

```
******************************************************
*** Run psi_fix.en_cl_fix_pkg_tb
******************************************************
Running Simulation
ghdl --elab-run --ieee=synopsys --std=08 -frelaxed-rules -Wno-shared --work=psi_fix en_cl_fix_pkg_tb --ieee-asserts=disable
*** cl_fix_width ***
*** cl_fix_from_real ***
*** cl_fix_to_real ***
*** cl_fix_from_bits_as_int ***
*** cl_fix_get_bits_as_int ***
*** cl_fix_resize ***
*** cl_fix_add ***
*** cl_fix_sub ***
*** cl_fix_mult ***
*** cl_fix_abs ***
*** cl_fix_neg ***
*** cl_fix_shift left ***
*** cl_fix_shift right  ***
*** cl_fix_max_value ***
*** cl_fix_min_value ***
*** cl_fix_max_real ***
/psi/psi_fpga_all/VHDL/en_cl_fix/vhdl/tb/en_cl_fix_pkg_tb.vhd:60:16:@0ms:(assertion error): ###ERROR### unsigned [expected: 3.75, got: 3.749999999999999]

```

**Comment by obruendl** [comment]
Should be fixed now. You have to use en_cl_fix 1.1.7.

**Comment by obruendl** [comment]
@rybaniec-r 

Could you confirm if the issue is fixed? And if so, would it be possible creating a new release based on the current stat of the develop branch? It contains quite some fixes that we actively use in our projects ...

... as you see, we actively use psi_fix at Enclustra :-)

**Comment by rybaniec-r** [comment]
@obruendl 
No the issue is still there. It looks like some floating-point operations are not precise in the GHDL. Still there are some changes required in en_cl_fix. I have prepared a workaround using a look-up table for power of 2 function. I also removed the real to integer casting so now it is possible to have a full precision when converting floats into fixed point numbers with more than 32 bits. It works nicely in both Modelsim and GHDL. I have pushed it here: https://github.com/paulscherrerinstitute/en_cl_fix/tree/bigreal
